src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs

```diff
@@ -0,0 +1,417 @@
+                            // to the thread that registered the target no longer
+                            // needing the stream (e.g. it was async awaiting and
+                            // was cancelled). Dispose the target to release any
+                            // resources it may have.
+                            target.Dispose();
+                        }
+                    }
+                    while (null != stream);
+                }
+
+                // Store the stream for when a target registers for the stream. If
+                // a target was already provided the stream, this will be null, thus
+                // representing that no existing stream is waiting to be consumed.
+                _stream = stream;
+            }
+            finally
+            {
+                // If a target took ownership of the semaphore, don't release it.
+                _streamSemaphore.Release();
+            }
+
+            // Dispose the previous stream if there was one.
+            previousStream?.Dispose();
+        }
+
+        protected virtual Stream RefreshStream()
+        {
+            return null;
+        }
+
+        private bool TestStream(Stream stream)
+        {
+            if (null == stream)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            if (stream is ExposedSocketNetworkStream networkStream)
+            {
+                // Update Connected state of socket by sending non-blocking zero-byte data.
+                Socket socket = networkStream.Socket;
+                bool blocking = socket.Blocking;
+                try
+                {
+                    socket.Blocking = false;
+                    socket.Send(Array.Empty<byte>(), 0, SocketFlags.None);
+                }
+                catch (Exception)
+                {
+                }
+                finally
+                {
+                    socket.Blocking = blocking;
+                }
+                return socket.Connected;
+            }
+            else if (stream is PipeStream pipeStream)
+            {
+                Debug.Assert(RuntimeInformation.IsOSPlatform(OSPlatform.Windows), "Pipe stream should only be used on Windows.");
+
+                // PeekNamedPipe will return false if the pipe is disconnected/broken.
+                return NativeMethods.PeekNamedPipe(
+                    pipeStream.SafePipeHandle,
+                    null,
+                    0,
+                    IntPtr.Zero,
+                    IntPtr.Zero,
+                    IntPtr.Zero);
+            }
+
+            return false;
+        }
+
+        private void RegisterTarget(StreamTarget target)
+        {
+            // Allow transport specific implementation to refresh
+            // the stream before possibly consuming it.
+            if (null == _stream)
+            {
+                _stream = RefreshStream();
```

Comment:
While I originally had PidEndpoint stateless as well, there were two issues that were brought up or I thought of:
1. In your comment at https://github.com/dotnet/diagnostics/pull/1303#discussion_r450620722, if the socket file is abandoned, then WaitForConnectionAsync will just keep trying to reconnect to the socket at that file. I wouldn't expect this method to return (or likely throw in this case), but would wait until there is a viable connection. That's why I brought in the non-blocking read/write stream testing that was originally used for the ReversedDiagnosticsServer.
2. In my earlier iterations and in your experimental changes, the WaitForConnectionAsync was creating the stream for that socket and then immediately disposing it. This will consume a connection from the runtime instance for each call into WaitForConnectionAsync. I assume the runtime instance handles this gracefully and will not diminish the number of simultaneous connections shortly after the stream is closed, however, it still feels like a wasted connection. Thus, I cached the stream so that it can be used for the next call to WaitForConnectionAsync method or just returned from the Connect method. But if disposing this stream for each WaitForConnectionAsync calls is okay, then keeping this class stateless should okay as well.